tools/lsp/common/rename_component.rs

```diff
@@ -516,41 +544,495 @@ pub fn rename_identifier_from_declaration(
+            )),
+            DeclarationNode::ExportName(name) => Ok(rename_local_export_name(
+                document_cache,
+                name,
+                &new_type,
+                local_rename_function(document_cache, name),
+            )),
+        }
+    }
+
+    #[cfg(test)]
+    fn as_declared_identifier(&self) -> syntax_nodes::DeclaredIdentifier {
+        match &self {
+            DeclarationNode::DeclaredIdentifier(identifier) => identifier.clone(),
+            _ => panic!("Wrong declaration node"),
+        }
+    }
+
+    #[cfg(test)]
+    fn as_internal_name(&self) -> syntax_nodes::InternalName {
+        match &self {
+            DeclarationNode::InternalName(name) => name.clone(),
+            _ => panic!("Wrong declaration node"),
+        }
+    }
+
+    #[cfg(test)]
+    fn as_export_name(&self) -> syntax_nodes::ExportName {
+        match &self {
+            DeclarationNode::ExportName(name) => name.clone(),
+            _ => panic!("Wrong declaration node"),
+        }
+    }
+}
+
+impl From<syntax_nodes::DeclaredIdentifier> for DeclarationNode {
+    fn from(value: syntax_nodes::DeclaredIdentifier) -> Self {
+        Self::DeclaredIdentifier(value)
+    }
+}
+
+impl From<syntax_nodes::InternalName> for DeclarationNode {
+    fn from(value: syntax_nodes::InternalName) -> Self {
+        Self::InternalName(value)
+    }
+}
+
+impl From<syntax_nodes::ExportName> for DeclarationNode {
+    fn from(value: syntax_nodes::ExportName) -> Self {
+        Self::ExportName(value)
+    }
+}
+
+impl TryFrom<SyntaxNode> for DeclarationNode {
+    type Error = String;
+
+    fn try_from(value: SyntaxNode) -> Result<Self, Self::Error> {
+        match value.kind() {
+            SyntaxKind::DeclaredIdentifier => Ok(Self::DeclaredIdentifier(value.into())),
+            SyntaxKind::InternalName => Ok(Self::InternalName(value.into())),
+            SyntaxKind::ExportName => Ok(Self::ExportName(value.into())),
+            _ => Err("Can not convert into an DeclarationNode".into()),
+        }
     }
 }
 
+fn find_last_declared_identifier_at_or_before(
+    token: SyntaxToken,
+    type_name: &SmolStr,
+) -> Option<syntax_nodes::DeclaredIdentifier> {
+    let mut token = Some(token);
+
+    while let Some(t) = token {
+        if t.kind() == SyntaxKind::Identifier {
+            let node = t.parent();
+            if node.kind() == SyntaxKind::DeclaredIdentifier
+                && i_slint_compiler::parser::identifier_text(&node).as_ref() == Some(type_name)
+            {
+                return Some(node.into());
+            }
+        }
+        token = t.prev_token();
+    }
+
+    None
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum DeclarationNodeQueryKind {
+    Component,
+    Type,
+    Unknown,
+}
+
+#[derive(Clone, Debug)]
+struct DeclarationNodeQuery {
+    kind: DeclarationNodeQueryKind,
+    name: SmolStr,
+    node: SyntaxNode,
+}
+
+impl DeclarationNodeQuery {
+    fn new(node: SyntaxNode) -> Option<Self> {
```

Comment:
This seems like it is an incomplete duplication of `language::token_info::token_info()`
That function also do the lookup and tell you the exact type this is referring to.
So I think it would be a better candidate.

Reply:
Latest push contains the work to redo this using token_info.

I have an additional PR ready that adds renaming properties... but that is entirely untested so far and also needs to be integrated into the property editor. But this is extensible:-)